### PR TITLE
DS-4440 GDPR - Anonymize statistics feature

### PR DESCRIPTION
**Pull request by samuelcambien**
Fixes #7777

This PR adds a script to anonymise records in the Solr statistics core. The script will anonymise the IP values by rewriting (‘masking’) the last part. This mask is configurable, both for ipv4 and ipv6 addresses.

For IPv4 addresses, the last number will be replaced by the mask, defined by the configuration key ‘anonymise_statistics.ip_v4_mask’ which defaults to ‘254’.
For example, 109.74.16.171 is rewritten as 109.74.16.254

For IPv6 address, the last two numbers will be replaced by the mask, defined by the configuration key ‘anonymise_statistics.ip_v6_mask’ which defaults to ‘FFFF:FFFF’
For example, 2001:0db8:85a3:0000:0000:8a2e:0370:7334 is rewritten as 2001:0db8:85a3:0000:0000:8a2e:FFFF:FFFF

For each anonymised record, the DNS field is also replaced by “anonymised”.

The program only processes records older than 90 days. This period can be altered with the config ‘anonymise_statistics.time_limit’ (expressed in days).

The command ‘anonymise-statistics’ has been added to launcher.xml to launch the script.

The script takes an optional parameter ‘-s [sleep]’ (expressed in ms), which will make the Java thread sleep between the calls to Solr to reduce the load impact.

The Solr service commit mechanism is also optimised by adding multi-threading support. The script takes an optional parameter ‘-t [threads]’ to indicate how many threads the Solr service can use for this, if not given the thread count defaults to 2.

Besides the 'anonymise-statistics' script, which anonymises existing statistics record, the PR also allows for statistics records to be anonymised the moment they are created.
Enabling this feature can be done by setting the configuration parameter "anonymise_statistics.anonymise_on_log" to true.
When this configuration property is not set, the feature is disabled by default.

Finally, the PR also adds pagination support to SolrLoggerService

The work done on this PR is sponsored and led by the Humboldt-Universität zu Berlin.

### CODE REVIEW

**Comment by samuelcambien** [comment]
https://jira.lyrasis.org/browse/DS-4440

**Comment by tdonohue** [comment]
@samuelcambien : Thanks for this! I took a quick glance at the code and I'd like to ask that you please add Javadocs & Unit (or Integration) Tests for all new methods/classes.  Our policies for `master` are to require those.  I'll gladly give this another review once JavaDocs & tests are added.

**Comment by tdonohue** [comment]
@samuelcambien : Thanks for this! I took a quick glance at the code and I'd like to ask that you please add Javadocs & Unit (or Integration) Tests for all new methods/classes.  Our policies for `master` are to require those.  I'll gladly give this another review once JavaDocs & tests are added.

File: dspace/config/launcher.xml
Comment (pnbecker) [inline dspace/config/launcher.xml]: This looks like a typo to me. Shouldn't that be `org.dspace.statistics.AnonymizeStatistics`?

File: dspace-api/src/main/java/org/dspace/statistics/AnonymizeStatistics.java
Comment (pnbecker) [inline dspace-api/src/main/java/org/dspace/statistics/AnonymizeStatistics.java]: Just a suggestion, not a change request: I would expect a negative exit code here. If the help parameter is given, all other parameters are ignored and the task stopped without having done anything. That doesn't seem to be a successful run to me.

File: dspace-api/src/main/java/org/dspace/statistics/AnonymizeStatistics.java
Comment (pnbecker) [inline dspace-api/src/main/java/org/dspace/statistics/AnonymizeStatistics.java]: suggestion: rephrase to "Sleep a certain time **given in milliseconds** between each solr request"

**Review by pnbecker — CHANGES_REQUESTED** [review]
I have one really simple change request and two suggestions (see inline comments). Overall this looks good to me. Once the launcher.xml is fixed, I'm +1 on this PR.

It will need some small additions to the documentation once it is merged.

File: dspace/config/launcher.xml
Comment (ssolim) [reply dspace/config/launcher.xml]: Yes this is a typo, this needs to be fixed.

File: dspace-api/src/main/java/org/dspace/statistics/AnonymizeStatistics.java
Comment (ssolim) [reply dspace-api/src/main/java/org/dspace/statistics/AnonymizeStatistics.java]: +1 for this suggestion

**Comment by ssolim** [comment]
This is the port to master, https://github.com/DSpace/DSpace/pull/2693 to dspace-6x

**Comment by abollini** [comment]
I'm ok with the feature behind this PR and I agree to include it in the 6.4 (and 5.11). 
The only question that I have is if it was not easier to use the SOLR buildin facilities instead than provide our own java script. We can do that exporting the statistics document to a CSV file, make the required changes and importing them back. This would be probably easier and much more lightweight for the server than update each document individually

**Comment by mwoodiupui** [comment]
What tool, already available at every DSpace site, would we use to automate "make the required changes"?

**Comment by ssolim** [comment]
The anonymiser script has threads and bulk-commit options. I can give you here an idea of the performance: bulk size: 5000 threads 4 ~ 250k documents per hour. On a nightly cronjob with  around 50k stats records the script runs 10 minutes every day. Ofc these numbers depends on system parameters.
For a large repository with a lot accesses , the "anonymise_statistics.anonymise_on_log" feature can be used.
I cannot say how this compares to SOLR buildin facilities but i think this feature is well embedded into the standard dspace environment and easy to integrate...

File: dspace/config/launcher.xml
Comment (samuelcambien) [reply dspace/config/launcher.xml]: You are correct, fixed this!

File: dspace-api/src/main/java/org/dspace/statistics/AnonymizeStatistics.java
Comment (samuelcambien) [reply dspace-api/src/main/java/org/dspace/statistics/AnonymizeStatistics.java]: Good suggestion, I've changed it.

File: dspace-api/src/main/java/org/dspace/statistics/AnonymizeStatistics.java
Comment (samuelcambien) [reply dspace-api/src/main/java/org/dspace/statistics/AnonymizeStatistics.java]: Good suggestion, I've changed it.

**Comment by benbosman** [comment]
I don't see any problems with this approach, nor with the solution to take advantage of solr's atomic updates which is optimized for these small changes to existing documents

**Comment by paulo-graca** [comment]
I like the approach of having a configuration to control whether anonymization is active or not.

**Comment by abollini** [comment]
Ok, as the performance looks good I think that the benefit of a deeper integration are higher than the cost. I would like to have a test in place for that that verify that all the statistics documents are keep in the index after the update and no unexpected changes to other fields happen. Use of automatic update require a proper configuration in the statistics schema, I haven't checked if we already meet them or not https://lucene.apache.org/solr/guide/7_7/updating-parts-of-documents.html#field-storage
but it will be useful to add a comment in the schema to note that.

**Comment by AlexanderS** [comment]
I do not think, that masking the last 4 octets of an IPv6 address is sufficient.

RIPE has some recommendations for IPv6 assignment for end-users: https://www.ripe.net/publications/docs/ripe-690
> It is strongly discouraged to assign prefixes longer than /56 unless there are very strong and unsolvable technical reasons for doing this.
 
> Assigning a /64 or longer prefix does not conform to IPv6 standards and will break functionality in customer LANs.

So the users are sill uniquely identifiable after masking out the last 32 bits.

**Comment by kshepherd** [comment]
@samuelcambien is it hard to increase the masking for ipv6 as per @AlexanderS 's comment?
I probably wouldn't treat this as a blocker as it is an improvement on existing, and afaik stats are still most likely to be logged as ipv4 addresses, but it would be nice to increase the masking if possible

**Review by pnbecker — APPROVED** [review]
My issues were resolved. I agree with Alexander that we should mask more of the IPv6 address and I agree with Kim, that this shouldn't block this from being merged. Therefore: +1

@benbosman @samuelcambien Would you expect performance issues if we change the regex and the configuration that are responsible for masking the IPv6 address? Do I see it correctly that these two lines would be needed to change if we would like to mask more from the IPv6 address?

**Comment by AlexanderS** [comment]
One additional thing I just noticed: Can we get a config file with the available options and a short documentation? Maybe in `dspace/config/modules/anonymise_statistics.cfg` or something like this?

**Comment by kshepherd** [comment]
Hi @AlexanderS - indeed, if you have a look at the 6.x version of this PR, we did discuss that and it's been addressed, does it look OK to you? It should be easy to port those recent improvements back over to this master PR: #2693

**Comment by kshepherd** [comment]
Hi @samuelcambien , we just need to fix the checkstyle for the latest commits. The last travis build failed with:
```
[INFO] 

[INFO] --- maven-checkstyle-plugin:3.1.0:check (verify-style) @ dspace-api ---

[INFO] There is 1 error reported by Checkstyle 8.30 with /home/travis/build/DSpace/DSpace/dspace-api/../checkstyle.xml ruleset.

[ERROR] src/main/java/org/dspace/statistics/AnonymizeStatistics.java:[17] (imports) CustomImportOrder: Extra separation in import group before 'org.apache.commons.cli.Option.builder'
```

**Comment by samuelcambien** [comment]
Hi kshepherd, I fixed the checkstyle issue, sorry for the delay!

**Comment by tdonohue** [comment]
@samuelcambien or @kshepherd : I'm uncertain of the status of this PR as it pertains to what was merged for 6.4?  Is this PR updated with all the feedback from the 6.4 PR?

It also looks like this has already had general approval from @paulo-graca and possibly also @abollini and @benbosman .  But, it looks like it's still officially at just +1.  Have others tested this out enough to formally add another +1 here?  

Generally, I'm trying to figure out why this PR seems "stalled" since it looks like it was approved, tested & merged already for the upcoming 6.4 release.

**Comment by kshepherd** [comment]
@tdonohue @samuelcambien yes, the improvements in #2693 (6.4) need to be ported as far as i know - i think they should be able to just be cherry-picked across

**Comment by tdonohue** [comment]
@samuelcambien : can you get this PR updated with all the fixes applied to 6.4?  If so, I can work on getting a second +1, so that we can get this merged & ensure it's also available in 7.0.  Thanks!

**Comment by samuelcambien** [comment]
Hi Tim and Kim, I might be wrong, but as far as I can see all commits from the 6.4 branch are in place in this branch.

**Comment by tdonohue** [comment]
@samuelcambien : My mistake, you are correct. This PR is already synced with #2693...I was thrown off as the dates look different...several of the May feedback changes to 2693 show up in this PR as being done in April, and I wrongly assumed there was outstanding work to be done on this PR.

In any case, this one just needs a final +1 from someone before we get it merged.  If one of the past reviewers (@kshepherd , @AlexanderS , @paulo-graca or @abollini ) is interested in re-reviewing, we could likely get this merged quickly. Otherwise, I'll look for time on my schedule to help this along.

File: dspace-api/src/main/java/org/dspace/statistics/AnonymizeStatistics.java
Comment (tdonohue) [inline dspace-api/src/main/java/org/dspace/statistics/AnonymizeStatistics.java:49]: Please add JavaDocs to this class to document it

File: dspace-api/src/main/java/org/dspace/statistics/AnonymizeStatistics.java
Comment (tdonohue) [inline dspace-api/src/main/java/org/dspace/statistics/AnonymizeStatistics.java:172]: Even though this is a private method, it contains most of the logic of the class. So, it should have JavaDocs or inline comments to describe *how* it is anonymizing statistics, etc.

File: dspace-api/src/main/java/org/dspace/statistics/AnonymizeStatistics.java
Comment (tdonohue) [inline dspace-api/src/main/java/org/dspace/statistics/AnonymizeStatistics.java:240]: Please add JavaDocs to this internal class.

File: dspace-api/src/main/java/org/dspace/statistics/AnonymizeStatistics.java
Comment (tdonohue) [inline dspace-api/src/main/java/org/dspace/statistics/AnonymizeStatistics.java:221]: Could we describe this method with JavaDocs *or* add an inline comment above the query to describe *which* documents it's getting.  It seems like it's returning all docs from the last 90 days (or the configured time period)? Is that correct?

File: dspace-api/src/main/java/org/dspace/statistics/SolrLoggerServiceImpl.java
Comment (tdonohue) [inline dspace-api/src/main/java/org/dspace/statistics/SolrLoggerServiceImpl.java:881]: Could you add some inline comments regarding the shard logic here?  It's not clear at a glance why we are removing the version and shard fields, and then adding the entire solrDocument back to the shard.

File: dspace-api/src/main/java/org/dspace/statistics/SolrLoggerServiceImpl.java
Comment (tdonohue) [inline dspace-api/src/main/java/org/dspace/statistics/SolrLoggerServiceImpl.java:1688]: New public method should have JavaDocs

File: dspace-api/src/main/java/org/dspace/statistics/service/SolrLoggerService.java
Comment (tdonohue) [inline dspace-api/src/main/java/org/dspace/statistics/service/SolrLoggerService.java:120]: Please add JavaDocs to this new method

File: dspace-api/src/main/java/org/dspace/statistics/service/SolrLoggerService.java
Comment (tdonohue) [inline dspace-api/src/main/java/org/dspace/statistics/service/SolrLoggerService.java:183]: Please add JavaDocs to this new method

File: dspace-api/src/main/java/org/dspace/statistics/service/SolrLoggerService.java
Comment (tdonohue) [inline dspace-api/src/main/java/org/dspace/statistics/service/SolrLoggerService.java:189]: Please add JavaDocs to this new method

File: dspace-api/src/main/java/org/dspace/statistics/service/SolrLoggerService.java
Comment (tdonohue) [inline dspace-api/src/main/java/org/dspace/statistics/service/SolrLoggerService.java:222]: Please add JavaDocs to this new method

**Review by tdonohue — COMMENTED** [review]
@samuelcambien : In an effort to help move this along, I've given this a code review (haven't tested it yet).  Overall the code looks reasonable, but I've noticed that it doesn't include any JavaDocs or inline comments.  So, I'd like to see JavaDocs added & a few areas could use inline comments to describe the logic.  See comments inline below.

**Comment by tdonohue** [comment]
@samuelcambien : What Codecov is noting in the "patch" report is that this PR looks to be missing tests altogether (as the "patch" report only analyzes the test coverage of the *new code* added in the PR).  

That said, the PR still manages to pass the "project" report because the PR is small (so the net decrease on the overall code coverage is minimal).

However, I completely overlooked that this doesn't have any unit or integration tests (which as we know are required in DSpace 7). To be fair, this is an area that doesn't have many unit tests at all (in the dspace-api).   But, I wonder if we could _minimally_ find a way to add an Integration Test (or two) to the REST API (`StatisticsRestRepositoryIT`) to show that anonymization doesn't break statistical reports?  I admit, I haven't tested this myself...but I'm hoping that running this script does *not* break the current usage reports (which are used by your other PR (https://github.com/DSpace/dspace-angular/pull/851).

Overall, the code here looks good. It has a lot of general approval, but the Codecov report shows that this has no tests.  So, I'd prefer, if possible, if we could minimally add a test or two to show this is working as expected.

**Comment by tdonohue** [comment]
@samuelcambien : Could I ask why you've removed shard support in the last commit here?  It seems like, in order to fully anonymize statistics, we'd need to also anonymize the shards. Am I misunderstanding, or maybe you are now taking a different approach?

In yesterday's meeting, @artlowel noted that you had hit issues with providing *tests* which prove the shard support works correctly.  If we cannot automate the testing of the shard support, I'm OK with us testing that more manually...provided that we can create unit/integration tests which prove anonymization works properly in the main solr core.

**Comment by samuelcambien** [comment]
Hi Tim,

I found out that with the code as it was, not only the tests were failing but the functionality itself didn't work either.
When I tested with a non-sharded core, [the '[shard]' field](https://github.com/DSpace/DSpace/pull/2692/commits/02d51dcb58ffc3751d738286f501504ea17810f2#diff-da2eb9c825f6877a3e37f78b99981e90714d47c664fb2c712e0ad10f4a487c58L207) doesn't contain the main solr endpoint anymore like it did in DSpace 6, but now has the value '[not a shard request]', making the code crash there.

This means another strategy has to be worked out to work with sharded cores.
I've tried to set up a sharded core by using the 'dspace stats-util --shard' command, but this didn't work: I get an error response on the [yearRangeQuery] (https://github.com/DSpace/DSpace/blob/main/dspace-api/src/main/java/org/dspace/statistics/SolrLoggerServiceImpl.java#L1172).
Because of this, I thought I'd make sure the code works for non-sharded cores, and to look into sharding support when this is fixed.

It also seems that, with solr 7, sharding is only supported in 'SolrCloud mode':
I tried to shard my local statistics core with the SPLITSHARD command, as described in the [docs](https://lucene.apache.org/solr/guide/8_2/shard-management.html), resulting in the error message 'Solr instance is not running in SolrCloud mode.'

**Comment by tdonohue** [comment]
Thanks @samuelcambien for the further explanation -- I was just wanting to understand the decisions behind this change, and it's now clearer to me. It looks like this is now passing tests.  Does that mean it's ready for a re-review now?

**Comment by samuelcambien** [comment]
You're welcome! Sorry forgot to mention: I've added a general test on the anonymization to StatisticsRestRepositoryIT, so yes if you could take a look whether this is sufficient?

File: dspace-server-webapp/src/test/java/org/dspace/app/rest/StatisticsRestRepositoryIT.java
Comment (tdonohue) [inline dspace-server-webapp/src/test/java/org/dspace/app/rest/StatisticsRestRepositoryIT.java:1345]: This is a good test method overall...but it's *very* difficult to tell here that the IP address _is actually_ anonymized.  Could you add a comment here to note that the IP was anonymized to end in 255?  or you could use a variable for the first 3 parts of the IP address to show only the *ending* changes.  

It literally took me 5 mins to realize that this IP *is* anonymized.  I originally started writing a comment here to ask why this isn't anonymized at all. :) 

So, I think these tests are good enough to prove things are working. They just could use a few inline comments (or use of variables) to better describe how the anonymization is occurring, so that they are easier for others to maintain & understand.

**Review by tdonohue — CHANGES_REQUESTED** [review]
@samuelcambien : This is looking better to me.  A few minor comments inline from my latest code review.

@pnbecker : Any chance you could give this another review or test?  It's received a lot of updates since your +1 back in April.  So, it'd be good to verify that you still vote +1 here, otherwise I'll look for another reviewer.

**Comment by samuelcambien** [comment]
Hi Tim, those are two valid remarks indeed, addressed them in the last two commits.

**Comment by samuelcambien** [comment]
There's a test failure but I have no clue what's going wrong, it seems unrelated to my changes... @tdonohue, or someone else, do you happen to know this error? Thanks

**Comment by tdonohue** [comment]
@samuelcambien : you are correct...the unit test failure is unrelated to your PR changes.  It was also discovered in another PR, and I described why in this comment: https://github.com/DSpace/DSpace/pull/2746#issuecomment-738897905

I'll work on getting the fix into `main`, so that you can simply rebase this PR to fix the problem.

**Comment by samuelcambien** [comment]
Ok, success!

**Comment by tdonohue** [comment]
@samuelcambien : I've just merged #2746 which includes a fix for the Unit test failure you were seeing. If you rebase on `main`, that issue should now be gone.  Thanks!

**Review by tdonohue — APPROVED** [review]
👍 @samuelcambien : Finally got back to this today.  The changes I previously requested were all made.  I also tested this and found it to be successful with IPv4 (which is what I have available).  I had to ensure my `usage-statistics.dbfile` setting was set appropriately. I already had some statistics in my `statistics` core, so I just had to decrease the default `anonymize_statistics.time_threshold` to a few days.  Then, after running `./dspace anonymize-statistics`  I was able to see in my `statistics` core that the `ip` was anonymized to always end in 255.  Thanks!  This looks good to me.

**Comment by KevinVdV** [comment]
Did a review, found one small thing that we might want to fix:

There are new tests added to the "StatisticsRestRepositoryIT" in the dspace-server-webapp, but these tests don't have anything to do with the server webapp. These should be moved to the dspace-api package.

**Comment by samuelcambien** [comment]
Hi Kevin,
Thanks for your feedback!
I've moved the tests to dspace-api.

**Review by KevinVdV — APPROVED** [review]
The tests are now in the dspace-api. Approving.

**Review by tdonohue — APPROVED** [review]
@samuelcambien : Thanks for your continued hard work on this!  I'm merging this now as it's at +2 and the most recent minor tweaks look good to me.

When you get the chance, would you be willing to add some basic documentation about the new Anonymization settings in `usage-statistics.cfg` to our DSpace 7 docs at: https://wiki.lyrasis.org/display/DSDOC7x/SOLR+Statistics ?  Thanks!

**Pull request merged**
